plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/models/OpenShiftProjectUIModel.java

```diff
@@ -90,48 +91,85 @@ public IProject getProject() {
 		return this.project;
 	}
 
-
 	@Override
 	public IResource getResource() {
 		return getProject();
 	}
 
-	public boolean isLoading() {
-		return false;
+	@Override
+	public IOpenShiftConnection getConnection() {
+		return this.connection;
 	}
 	
+	@Override
+	public String toString() {
+		return this.project.getName();
+	}
+
+	@Override
+	public Object getParent() {
+		return null;
+	}
+
 	@SuppressWarnings("unchecked")
 	@Override
-	public <T extends IResource> void setResources(Collection<T> resources, String kind) {
-		switch(kind) {
-		case ResourceKind.BUILD:
-			setBuildResources((Collection<IBuild>) resources);
-			break;
-		case ResourceKind.BUILD_CONFIG:
-			setBuildConfigResources((Collection<IResource>) resources);
-			break;
-		case ResourceKind.DEPLOYMENT_CONFIG:
-			setDeploymentConfigResources((Collection<IResource>) resources);
-			break;
-		case ResourceKind.IMAGE_STREAM:
-			setImageStreamResources((Collection<IResource>) resources);
-			break;
-		case ResourceKind.POD:
-			setPodResources((Collection<IPod>) resources);
-			break;
-		case ResourceKind.ROUTE:
-			setRouteResources((Collection<IResource>) resources);
-			break;
-		case ResourceKind.REPLICATION_CONTROLLER:
-			setReplicationControllerResources((Collection<IResource>) resources);
-			break;
-		case ResourceKind.SERVICE:
-			setServiceResources((Collection<IResource>) resources);
-			break;
-		default:
+	public <T> T getAdapter(final Class<T> adapter) {
+		if(adapter.equals(IProject.class)) {
+			return (T) this.project;
+		} else if(adapter.equals(IResource.class)) {
+				return (T) this.project;
+		} else if(adapter.equals(Connection.class)) {
+			return (T) this.connection;
 		}
+		return null;
+	}
+
+	@Override
+	public Collection<IResourceUIModel> getBuilds() {
+		return this.mapper.getDeployments().stream().flatMap(deployment -> deployment.getBuilds().stream())
+				.collect(Collectors.toList());
+	}
+
+	@Override
+	public Collection<IResourceUIModel> getImageStreams() {
+		return this.mapper.getDeployments().stream().flatMap(deployment -> deployment.getImageStreams().stream())
```

Comment:
This may be not thread safe. Yes, getImageStreams() returns synchronized list, but iterator is not thread safe, and java docs say that operations with iteration should be manually synchronized. The same is true for all other similar getters.